File: client/dotnet-client/NeoApi/Validation/Validators.cs
using Kotak.Neo.Exceptions;

namespace Kotak.Neo.Validation;

public static class Validators
{
    private static void NonEmpty(string? value, string name)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ApiValueError($"{name} is mandatory");
    }

    public static void ValidatePlaceOrder(string exchangeSegment, string product, string price,
        string orderType, string quantity, string validity, string tradingSymbol, string transactionType)
    {
        NonEmpty(exchangeSegment, "exchange_segment");
        NonEmpty(product, "product");
        NonEmpty(price, "price");
        NonEmpty(orderType, "order_type");
        NonEmpty(quantity, "quantity");
        NonEmpty(validity, "validity");
        NonEmpty(tradingSymbol, "trading_symbol");
        NonEmpty(transactionType, "transaction_type");
        if (!Settings.ExchangeSegment.ContainsKey(exchangeSegment))
            throw new ApiValueError($"invalid exchange_segment: {exchangeSegment}");
        if (!Settings.Product.ContainsKey(product))
            throw new ApiValueError($"invalid product: {product}");
        if (!Settings.OrderType.ContainsKey(orderType))
            throw new ApiValueError($"invalid order_type: {orderType}");
        var tt = transactionType.ToUpperInvariant();
        if (tt != "B" && tt != "S" && tt != "BUY" && tt != "SELL")
            throw new ApiValueError("transaction_type must be B/S");
    }

    public static void ValidateCancelOrder(string orderId) => NonEmpty(orderId, "order_id");
    public static void ValidateOrderHistory(string orderId) => NonEmpty(orderId, "order_id");

    public static void ValidateMargin(string exchangeSegment, string price, string orderType,
        string product, string quantity, string instrumentToken, string transactionType)
    {
        NonEmpty(exchangeSegment, "exchange_segment");
        NonEmpty(price, "price");
        NonEmpty(orderType, "order_type");
        NonEmpty(product, "product");
        NonEmpty(quantity, "quantity");
        NonEmpty(instrumentToken, "instrument_token");
        NonEmpty(transactionType, "transaction_type");
        if (!Settings.ExchangeSegment.ContainsKey(exchangeSegment))
            throw new ApiValueError($"invalid exchange_segment: {exchangeSegment}");
    }

    public static void ValidateLimits(string segment, string exchange, string product)
    {
        if (!Settings.SegmentLimits.Contains(segment))
            throw new ApiValueError($"invalid segment: {segment}");
        if (!Settings.ExchangeLimits.Contains(exchange))
            throw new ApiValueError($"invalid exchange: {exchange}");
        if (!Settings.ProductLimits.Contains(product))
            throw new ApiValueError($"invalid product: {product}");
    }
}

File: client/java-client/src/main/java/com/kotak/neo/client/validation/Validators.java
package com.kotak.neo.client.validation;

import com.kotak.neo.client.Settings;
import com.kotak.neo.client.exceptions.ApiValueError;

public final class Validators {
    private Validators() {}

    private static void nonEmpty(String val, String name) {
        if (val == null || val.trim().isEmpty())
            throw new ApiValueError(name + " is mandatory");
    }

    public static void validatePlaceOrder(String exchangeSegment, String product, String price,
                                          String orderType, String quantity, String validity,
                                          String tradingSymbol, String transactionType) {
        nonEmpty(exchangeSegment, "exchange_segment");
        nonEmpty(product, "product");
        nonEmpty(price, "price");
        nonEmpty(orderType, "order_type");
        nonEmpty(quantity, "quantity");
        nonEmpty(validity, "validity");
        nonEmpty(tradingSymbol, "trading_symbol");
        nonEmpty(transactionType, "transaction_type");
        if (!Settings.EXCHANGE_SEGMENT.containsKey(exchangeSegment))
            throw new ApiValueError("invalid exchange_segment: " + exchangeSegment);
        if (!Settings.PRODUCT.containsKey(product))
            throw new ApiValueError("invalid product: " + product);
        if (!Settings.ORDER_TYPE.containsKey(orderType))
            throw new ApiValueError("invalid order_type: " + orderType);
        String tt = transactionType.toUpperCase();
        if (!tt.equals("B") && !tt.equals("S") && !tt.equals("BUY") && !tt.equals("SELL"))
            throw new ApiValueError("transaction_type must be B/S");
    }

    public static void validateCancelOrder(String orderId) { nonEmpty(orderId, "order_id"); }
    public static void validateOrderHistory(String orderId) { nonEmpty(orderId, "order_id"); }

    public static void validateMargin(String exchangeSegment, String price, String orderType,
                                      String product, String quantity, String instrumentToken,
                                      String transactionType) {
        nonEmpty(exchangeSegment, "exchange_segment");
        nonEmpty(price, "price");
        nonEmpty(orderType, "order_type");
        nonEmpty(product, "product");
        nonEmpty(quantity, "quantity");
        nonEmpty(instrumentToken, "instrument_token");
        nonEmpty(transactionType, "transaction_type");
        if (!Settings.EXCHANGE_SEGMENT.containsKey(exchangeSegment))
            throw new ApiValueError("invalid exchange_segment: " + exchangeSegment);
    }

    public static void validateLimits(String segment, String exchange, String product) {
        if (!Settings.SEGMENT_LIMITS.contains(segment))
            throw new ApiValueError("invalid segment: " + segment);
        if (!Settings.EXCHANGE_LIMITS.contains(exchange))
            throw new ApiValueError("invalid exchange: " + exchange);
        if (!Settings.PRODUCT_LIMITS.contains(product))
            throw new ApiValueError("invalid product: " + product);
    }
}
